### ISSUE TITLE
Autogenerated swagger.json Comes with Errors as per OpenApi 3 spec

### ISSUE THREAD

**Issue by luli0822** [lukeautry/tsoa#389]
Hello,

I am using tsoa that is autogenerating swagger.json. But it failed because of some errors in the browser console. I tried it in [editor.swagger.io](https://editor.swagger.io/), and it reported like the screenshot attached here. 
Am I missing something somewhere?
Also, how do we verify api with headers and params, e.g. Authorization, and others?
Thanks and appreciate any advice and feedbacks.

![image](https://user-images.githubusercontent.com/11763452/62630582-365fde80-b8fd-11e9-8c3f-a75afb53c6e0.png)

The __swagger.json__ is:
```
    {

    "components": {
        "examples": {},
        "headers": {},
        "parameters": {},
        "requestBodies": {},
        "responses": {},
        "schemas": {
            "ApiResponse": {
                "description": "Swagger API response",
                "properties": {
                    "message": {
                        "type": "string"
                    },
                    "status": {
                        "type": "object"
                    },
                    "data": {
                        "type": "object",
                        "nullable": true
                    }
                },
                "required": [
                    "message",
                    "status"
                ],
                "type": "object"
            },

            "UserResponse": {
                "properties": {
                    "data": {
                        "type": "object"
                    },
                    "link": {
                        "type": "string",
                        "nullable": true
                    }
                },
                "required": [
                    "data"
                ],
                "type": "object"
            },

            "ContactRequest": {
                "description": "Request body to create Contact",
                "properties": {
                    "firstName": {
                        "type": "string"
                    },
                    "lastName": {
                        "type": "string"
                    },
                    "email": {
                        "type": "string",
                        "nullable": true
                    },
                    "phoneNumber": {
                        "type": "string"
                    }
                },
                "required": [
                    "firstName",
                    "lastName",
                    "phoneNumber"
                ],
                "type": "object"
            },

            "AddressRequest": {
                "properties": {
                    "addressLine1": {
                        "type": "string"
                    },
                    "addressLine2": {
                        "type": "string",
                        "nullable": true
                    },
                    "province": {
                        "type": "string"
                    },
                    "city": {
                        "type": "string"
                    },
                    "country": {
                        "type": "string"
                    },
                    "postCode": {
                        "type": "string"
                    }
                },
                "required": [
                    "addressLine1",
                    "province",
                    "city",
                    "country",
                    "postCode"
                ],
                "type": "object"
            },

            "EntityType": {
                "enum": [
                    "Person",
                    "Organization"
                ],
                "type": "string"
            },

            "RegistrationType": {
                "enum": [
                    "admin",
                    "claim registrant",
                    "claimant",
                    "3"
                ],
                "type": "string"
            },

            "UserRequest": {
                "description": "Request body to register a new user\nwith Contact and Address info.",
                "properties": {
                    "userName": {
                        "type": "string"
                    },
                    "phoneNumber": {
                        "type": "string"
                    },
                    "password": {
                        "type": "string"
                    },
                    "contact": {
                        "$ref": "#/components/schemas/ContactRequest"
                    },
                    "address": {
                        "$ref": "#/components/schemas/AddressRequest"
                    },
                    "type": {
                        "$ref": "#/components/schemas/EntityType"
                    },
                    "role": {
                        "$ref": "#/components/schemas/RegistrationType"
                    }
                },
                "required": [
                    "userName",
                    "phoneNumber",
                    "password",
                    "contact",
                    "address",
                    "type",
                    "role"
                ],
                "type": "object"
            }
        },

        "securitySchemes": {
            "api_key": {
                "type": "apiKey",
                "name": "access_token",
                "in": "header"
            }
        }
    },

    "info": {
        "title": "Attribution Ledger OpenApi",
        "version": "0.1.5",
        "description": "This is my code repository",
        "license": {
            "name": "ISC"
        }
    },

    "openapi": "3.0.0",

    "paths": {
        "/users/reset-password": {
            "post": {
                "operationId": "ResetPassword",
                "produces": [
                    "application/json"
                ],
                "responses": {
                    "202": {
                        "description": "Reset accepted"
                    },
                    "400": {
                        "description": "Bad request",
                        "schema": {
                            "$ref": "#/components/schemas/ApiResponse"
                        }
                    },
                    "401": {
                        "description": "Unauthorized request",
                        "schema": {
                            "$ref": "#/components/schemas/ApiResponse"
                        }
                    },
                    "404": {
                        "description": "Registration not found",
                        "schema": {
                            "$ref": "#/components/schemas/ApiResponse"
                        }
                    }
                },
                "description": "The REST endpoint to reset password",
                "tags": [
                    "UserController"
                ],
                "security": [
                    {
                        "jwt": []
                    }
                ],
                "parameters": [],
                "requestBody": {
                    "content": {
                        "application/json": {
                            "schema": {
                                "type": "object"
                            }
                        }
                    }
                }
            }
        },
        "/users": {
            "post": {
                "operationId": "Register",
                "produces": [
                    "application/json"
                ],
                "responses": {
                    "201": {
                        "description": "Created",
                        "schema": {
                            "$ref": "#/components/schemas/UserResponse"
                        }
                    },
                    "409": {
                        "description": "Conflict request",
                        "schema": {
                            "$ref": "#/components/schemas/ApiResponse"
                        }
                    },
                    "500": {
                        "description": "System error",
                        "schema": {
                            "$ref": "#/components/schemas/ApiResponse"
                        }
                    }
                },
                "description": "The REST endpoint to register a user",
                "tags": [
                    "UserController"
                ],
                "security": [
                    {
                        "jwt": []
                    }
                ],
                "parameters": [
                    {
                        "in": "header",
                        "name": "Authorization",
                        "required": false,
                        "type": "string"
                    }
                ],
                "requestBody": {
                    "content": {
                        "application/json": {
                            "schema": {
                                "$ref": "#/components/schemas/UserRequest"
                            }
                        }
                    }
                }
            },
            "get": {
                "operationId": "GetUser",
                "produces": [
                    "application/json"
                ],
                "responses": {
                    "200": {
                        "description": "OK",
                        "schema": {
                            "$ref": "#/components/schemas/UserResponse"
                        }
                    },
                    "500": {
                        "description": "System failure",
                        "schema": {
                            "$ref": "#/components/schemas/ApiResponse"
                        }
                    }
                },
                "description": "Retrieve all users based on the input role",
                "tags": [
                    "UserController"
                ],
                "security": [
                    {
                        "jwt": []
                    }
                ],
                "parameters": [
                    {
                        "in": "query",
                        "name": "role",
                        "required": true,
                        "type": "string"
                    },
                    {
                        "in": "header",
                        "name": "Authorization",
                        "required": false,
                        "type": "string"
                    }
                ]
            }
        }
    },

    "servers": [
        {
            "url": "http://localhost:3001/api/v2"
        }
    ]
}

```

The __UserController.ts__ is:
```
    @Route('/users')
    @Tags('UserController')
    export class UserController extends Controller {

     private _userService: UserProfileService = new UserProfileService();

    /**
     * The REST endpoint to register a user
     * @param requestBody 
     */
    @Post()
    @Security('jwt')
    @Response<ApiResponse>('500', 'System error')
    @SuccessResponse('201', 'Created')
    public async register(@Body() requestBody: UserRequest, @Header('Authorization') authorization: string): Promise<UserResponse> {
        const request: UserRequest = requestBody;
        let user;
        try {
             user = await this._userService.registerUser(request);
             this.setStatus(201);
             return createUserResponse(user.id);
        } catch(error) {
            this.setStatus(500);
            throw new Error(`System error due to ${error.message}`);
        }
        
    }

    /**
     * Retrieve all users based on the input role
     * @param role 
     * @param authorization 
     */
    @Get()
    @Security('jwt')
    @Response<ApiResponse>('500', 'System failure')
    @SuccessResponse('200', 'OK')
    public async getUser(@Query() role: string, @Header('Authorization') authorization: string):  Promise<UserResponse> {
        let users;
        try {
            users = await this._userService.findAllUsers(role);
            this.setStatus(200);
            return createUserResponse(users);
        } catch(error) {
            this.setStatus(500);
            throw new Error(`System error due to ${JSON.stringify(error)}`);
        }
    }
  }
```

The __tsoa.json__ is:
```
{

    "swagger": {
        "outputDirectory": "./app",
        "entryFile": "./app/index.ts",
        "host": "localhost:3001",
        "basePath": "/api/v2",
        "securityDefinitions": {
            "api_key": {
                "type": "apiKey",
                "name": "access_token",
                "in": "header"
            }
        },
        "version": "0.1.5",
        "specVersion": 3,
        "name": "Attribution Ledger OpenApi",
        "schemes": ["http"]
    },

    "routes": {
        "basePath": "/api/v2",
        "entryFile": "./app/index.ts",
        "routesDir": "./app",
        "middleware": "express",
        "authenticationModule": "./app/authentication.ts"
    },

    "ignore": [
        "**/node_modules/**"
    ]

 }
```

**Comment by dgreene1** [comment]
We can look into this. Thank you for your thorough information on how to reproduce this. In the meantime, please consider using `"specVersion": 2` in your config. This appears to be a problem with OAS 3. @endor does anything pop out to you in this?

@luli0822 can you open a separate issue for this other question?
> Also, how do we verify api with headers and params, e.g. Authorization, and others?
Thanks and appreciate any advice and feedbacks.

Also, when you add code, please use three back-ticks instead of one. It's hard to read your JSON.

**Comment by luli0822** [comment]
> We can look into this. Thank you for your thorough information on how to reproduce this. In the meantime, please consider using `"specVersion": 2` in your config. This appears to be a problem with OAS 3. @endor does anything pop out to you in this?
> 
> @luli0822 can you open a separate issue for this other question?
> 
> > Also, how do we verify api with headers and params, e.g. Authorization, and others?
> > Thanks and appreciate any advice and feedbacks.
> 
> Also, when you add code, please use three back-ticks instead of one. It's hard to read your JSON.

I am sorry that I was using the embedded default icon <> on the tool bar. Thanks a lot for letting me know the trick. Have already updated using three back-ticks.

**Comment by luli0822** [comment]
> We can look into this. Thank you for your thorough information on how to reproduce this. In the meantime, please consider using `"specVersion": 2` in your config. This appears to be a problem with OAS 3. @endor does anything pop out to you in this?
> 
> @luli0822 can you open a separate issue for this other question?
> 
> > Also, how do we verify api with headers and params, e.g. Authorization, and others?
> > Thanks and appreciate any advice and feedbacks.
> 
> Also, when you add code, please use three back-ticks instead of one. It's hard to read your JSON.

**Comment by luli0822** [comment]
I am sorry about closing the issue by accident. Just wanted to confirm that it works pretty well on Swagger 2.

**Comment by dgreene1** [comment]
@luli0822 we believe that this issue has been solved (in https://github.com/lukeautry/tsoa/pull/401) with the release of tsoa version v2.4.7. Please upgrade tsoa within your repo and let us know if you're still having the issue.

**Comment by luli0822** [comment]
> @luli0822 we believe that this issue has been solved (in #401) with the release of tsoa version v2.4.7. Please upgrade tsoa within your repo and let us know if you're still having the issue.

@dgreene1 The validation errors from OpenAPI 3 are resolved. Thanks very much. But I found another issue #423.
